tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+/// drop implementations on your various layers will not get called when your
+/// program exits. This means that if you're using a buffered writer as the
+/// inner writer for the `FlameLayer` you're not guaranteed to see all the
+/// events that have been emitted in the file by default.
+///
+/// To ensure all data is flushed when the program exits, `FlameLayer` exposes
+/// the [`flush_on_drop`] function, which returns a [`FlushGuard`]. The `FlushGuard`
+/// will flush the writer when it is dropped. If necessary, it can also be used to manually
+/// flush the writer.
+///
+/// [`flush_on_drop`]: struct.FlameLayer.html#method.flush_on_drop
+/// [`FlushGuard`]: struct.FlushGuard.html
+#[derive(Debug)]
+pub struct FlameLayer<S, W>
+where
+    S: Subscriber,
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+    _inner: PhantomData<S>,
+}
+
+/// An RAII guard for managing flushing a global writer that is
+/// otherwise inaccessible.
+///
+/// This type is only needed when using
+/// `tracing::subscriber::set_global_default`, which prevents the drop
+/// implementation of layers from running when the program exits.
+#[must_use]
+#[derive(Debug)]
+pub struct FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    out: Arc<Mutex<W>>,
+}
+
+impl<S, W> FlameLayer<S, W>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+    W: Write + 'static,
+{
+    /// Return a new FlameLayer that outputs all folded stack samples to the
+    /// provided writer
+    pub fn new(writer: W) -> Self {
+        // Initialize the start used by all threads when initializing the
+        // LAST_EVENT when constructing the layer
+        START.with(|_| ());
+        Self {
+            out: Arc::new(Mutex::new(writer)),
+            _inner: PhantomData,
+        }
+    }
+
+    /// Return a `FlushGuard` which will flush the `FlameLayer`'s writer when
+    /// it is dropped, or when `flush` is manually invoked on the guard.
+    pub fn flush_on_drop(&self) -> FlushGuard<W> {
+        FlushGuard {
+            out: self.out.clone(),
+        }
+    }
+}
+
+impl<W> FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    /// Flush the internal writer of the `FlameLayer`, ensuring that all
+    /// intermediately buffered contents reach their destination.
+    pub fn flush(&self) -> Result<(), Error> {
+        Ok(self
+            .out
+            .lock()
+            .map_err(|_| Kind::LockError)
+            .map_err(Error)?
+            .flush()
+            .map_err(Kind::FlushFile)
+            .map_err(Error)?)
+    }
+}
+
+impl<W> Drop for FlushGuard<W>
+where
+    W: Write + 'static,
+{
+    fn drop(&mut self) {
+        match self.flush() {
+            Ok(_) => (),
+            Err(e) => e.report(),
+        }
+    }
+}
+
+impl<S> FlameLayer<S, BufWriter<File>>
+where
+    S: Subscriber + for<'span> LookupSpan<'span>,
+{
+    /// Construct a `FlameLayer` that outputs to a `BufWriter` to the given path, and a
```

Comment:
```suggestion
    /// Construcst a `FlameLayer` that outputs to a `BufWriter` to the given path, and a
```
also, maybe link to `BufWriter`?